---
title: "Identifying Gene Co-Expression Modules in the Developing Human Brain"
output:
  pdf_document: default
  word_document: default
---
**Course**: MED263, "Bioinformatics Applications to Human Disease"  
**Preparer**: Kevin Chau (kkchau@ucsd.edu) (https://github.com/kkchau/WeightedNetAnalysisTut)

# Introduction
Gene co-expression networks are highly informative in the context of biological processes as well as identifying risk genes through a guilty-by-association framework. These co-expression networks are constructed such that nodes represent the elements of interest, the genes, and edges represent correlations between the expression patterns of those genes. A straighforward approach to the construction of these gene co-expression networks would consist of calculating pairwise correlations, e.g. Pearson correlations, and applying a hard threshold such that edges only exists between nodes that have a correlation that surpasses the threshold. However, this approach comes with the caveat of lost information. For example, if two genes exhibit a correlation very close to, but not meeting, an established threshold, that potential edge is lost.

In this practical, we will be using R, specifically the Weighted Gene Co-Expression Network Analysis (WGCNA) package, to identify and characterize gene co-expression modules from human brain developmental transcriptome expression data. We will be using publicly available gene expression data from the BrianSpan Atlas, creating the networks with WGCNA in R, and characterizing these modules with ENRICHR.  
Note: For reference, the following code was executed in a Docker container allocated with 3 CPUs and 4096 MB of memory.

## Set-Up
We will be using WGCNA for the actual network construction; data will be organized into *SummarizedExperiment* objects for ease of use. All of the following packages should already be installed if running from the corresponding Docker container.
```{r}
setwd("/home/rstudio/work")
```

```{r}
# ggplot standard theming
theme.standard <- theme(text=element_text(size=15),
                   panel.grid.major=element_blank(),
                   panel.grid.minor=element_blank(),
                   panel.background=element_blank(),
                   axis.line=element_line())
```


```{r}
# install.packages(c('matrixStats', 'Hmisc', 'splines', 
#                    'foreach', 'doParallel', 'fastcluster',
#                    'dynamicTreeCut', 'survival', 'viridisLite',
#                    'ggplot2', 'enrichR')
# source("https://bioconductor.org/biocLite.R")
# biocLite(c('GO.db', 'preprocessCore', 'imput', 'WGCNA', 'SummarizedExperiment'))
```


## Libraries
```{r}
library(WGCNA)
library(SummarizedExperiment)
library(enrichR)
library(ggplot2)
library(gridExtra)
```
```{r}
### Optional:
# Enable parallel processing for computationally intensive tasks 
# (soft thresholding, topological overlap, etc.)
# Go to your Docker settings (Settings -> Advanced) and check how many CPUs 
# you would like to dedicate to the process
# Then, uncomment and run the following lines
library(doParallel)
allowWGCNAThreads()
```


# Data

## Initial Data Structuring
To begin the data analysis, we will first download and extract gene expression data from the BrainSpan Atlas.  
NOTE: If running from the Docker container, the data files will already be downloaded
```{r}
# url <- "http://www.brainspan.org/api/v2/well_known_file_download/267666525"
# utils::download.file(url, destfile="/home/rstudio/brainSpan.zip", mode='wb')
# utils::unzip("brainSpan.zip", exdir="/home/rstudio/brainSpan")
# file.remove("brainSpan.zip")
```

The downloaded files consist of a RPKM expression matrix, sample metadata, and row metadata, along with a readme file.

**Question 1**  
How were expressions compiled (what were the tools used for alignment/quantification) and what was the reference?

**Answer**  
Read alignment was performed with Tophat using Gencode v10 annotations; quantification was done with SAMtools and RSEQtools (this can be found in the whitepaper from https://brainspan.org)

Next, we want to package the relevant data and metadata together into a single SummarizedExperiment. This is done to protect the integrity of the data and safeguard value mappings, yielding a nice reference in case any downstream analysis goes wrong.
```{r}
expr <- read.csv("/home/rstudio/data/brainSpan/expression_matrix.csv", header=FALSE)[, -1]
coldata <- read.csv("/home/rstudio/data/brainSpan/columns_metadata.csv")[, -1]
rowdata <- read.csv("/home/rstudio/data/brainSpan/rows_metadata.csv")[, -1]

row.names(expr) <- rowdata$ensembl_gene_id
colnames(expr) <- as.character(apply(coldata, 1, 
                                     FUN=function(x) paste(x[["donor_name"]], 
                                                           x[["structure_acronym"]], 
                                                           sep="."
                                                           )
                                     )
                               )
se.expr <- SummarizedExperiment(assays=list(rpkm=as.matrix(expr)), 
                                rowData=rowdata, 
                                colData=coldata
                                )
se.expr
```
The se.expr SummarizedExperiment object is essentially a collection of matrices linked by the appropriate mappings; that is, the columns of the assays (in this case, the rpkm assay) correspond to the "colnames" of the SummarizedExperiment, which in turn have the attributes listed in the "colData" field. The same concept applies to the rownames and rowData.


# Methods and Results

## Data filtering

### Expression filtering
Now, we will filter this expression data for lowly expressed genes since these features are likely representative of noise in the dataset. Additionally, this will reduce the size of the data and make it more managable for the more computationally intensive steps to follow. We first plot the density plot of the expression distribution to identify a suitable cutoff value.
```{r}
hist(assays(se.expr)$rpkm, xlim=c(0, 10), breaks=100000)
```
Given this plot we select an expression cutoff of 5 RPKM in at least 90% of samples (this is a very strict threshold, but it is necessary to reduce the data to a manageable size for the default Docker memory settings).
```{r}
se.expr.filt <- se.expr[apply(assays(se.expr)$rpkm >= 5, 1, sum) >= 0.9*ncol(se.expr),]

dim(se.expr.filt)
object.size(se.expr.filt)
```

**Question 2**
Which five genes have the highest overall (average across samples) expression?

**Answer**
ENSG00000252229(20443.862), ENSG00000252197(19443.306), ENSG00000240831(11947.972), ENSG00000243172(9417.093), ENSG00000239935(8261.913)

## Soft thresholding
A stipulation of WGCNA network construction is that the data should satisfy the scale-free topology criteria. Scale-free networks are networks whose degree distributions (the number of neighbors per node) follows a power function, such that there are few nodes with a large number of neighbors whereas the rest of the nodes have few neighbors. This gives rise to "network hubs," and biologically this implies that there are fewer "vulnerable" genes such that their disruption results in the shutting down of a pathway, for example.  
This step consists of testing this topology on our data for different integer power levels. We want to select a power level that sufficiently satisfies the scale-free topology while still preserving connectivity.  
NOTE: This will take a while to run. If possible, enable parallel processing with allowWGCNAThreads().
```{r}
pwr <- c(seq(2, 8), seq(9, 21, by=3))
soft.thresh <- pickSoftThreshold(t(assays(se.expr.filt)$rpkm), 
                                 powerVector=pwr, 
                                 networkType="signed", 
                                 verbose=5
                                 )
```


We can then plot the R^2 fit to the scale-free topology of the different power levels as well as the mean connectivity.
```{r fig.width=12, fig.height=7}
sftplt <- (ggplot(soft.thresh$fitIndices, 
                  aes(x=soft.thresh$fitIndices$Power,
                      y=(-sign(soft.thresh$fitIndices$slope)
                         * soft.thresh$fitIndices$SFT.R.sq)
                      )
                  )
           + geom_text(label=soft.thresh$fitIndices$Power)
           + scale_y_continuous(breaks=c(-0.5, 0, 0.5, 0.85, 1), 
                                labels=c('-0.5', '0', '0.5', '0.85', '1')
                                )
           + geom_line(y=0.85, colour='red')
           + labs(x="Power", 
                  y="R^2 Scale-Free Topology Fit", 
                  title="Scale-Free Topology Fit"
                  )
           + theme.standard
           )
conplt <- (ggplot(soft.thresh$fitIndices, 
                  aes(x=soft.thresh$fitIndices$Power,
                      y=soft.thresh$fitIndices$mean.k)
                  )
           + geom_text(label=soft.thresh$fitIndices$Power)
           + labs(x="Power", 
                  y="Mean Connectivity", 
                  title="Mean Connectivity per Power Level"
                  )
           + theme.standard
           )
grid.arrange(sftplt, conplt, ncol=2)
```

**Question 3**
What should we select as the power level for this data set?

**Answer**
Some integer around 15. This is the point where we have sufficient correlation with the scale-free topology criterion. Choosing a greater value leads to decreased connectivity. This is also the value reported in *soft.thresh$powerEstimate*.

## Adjacencies and the Topological Overlap Matrix
WGCNA constructs networks by first creating a correlation adjacency matrix, where pairwise correlations between each feature (gene) is calculated. These correlations are then scaled by the power level reported by the soft thresholding function. A topological overlap matrix (TOM) is then the result of taking the information provided by the adjacency matrix, essentially correlation values, and incorporating topological similarities, like neighbors and distances between nodes.
```{r}
adj <- adjacency(t(assays(se.expr.filt)$rpkm), power=soft.thresh$powerEstimate)
tom <- TOMsimilarity(adj)
distom <- 1 - tom
```


## Clustering of the TOM
We can perform hierarchical clustering on this TOM to cluster genes together
```{r fig.width=12, fig.height=7}
geneTree <- hclust(as.dist(distom), method='average')
```
```{r fig.width=15, fig.height=10, fig.align='center'}
plot(geneTree, xlab="", sub="", main="Topological Overlap Clustering", labels=FALSE)
```

Using *cutreeDynamic* to dynamically cut the dendrogram, we can perform a preliminary identification of gene modules.
```{r}
modSize <- 20   # minimum module size
dynMods <- cutreeDynamic(dendro=geneTree, distM=distom, minClusterSize=modSize)
table(dynMods)
```

This result shows that we have identified 25 modules (module 0 represents genes that cannot be classified into another module). We can plot module assignment under the dendrogram.
```{r fig.align='center', fig.height=7, fig.width=12}
dynCol <- labels2colors(dynMods)
plotDendroAndColors(dendro=geneTree, colors=dynCol, dendroLabels=FALSE, 
                    main="Dynamically Cut TOM Dendrogram", groupLabels="Dynamic Colors"
                    )
```

**Question 4**
What is a likely cause of such a large grey module?

**Answer**
Due to such strict thresholding, many genes are now left with insufficient co-expression neighbors and thus do not make the clustering cutoff.

We can further refine the module identification by clustering highly correlated modules together. This is done based on the "module eigengenes," or representative vectors of each module that incorporates an average expression pattern of the module members. We will use a correlation of 0.8 as the merging threshold.
```{r fig.align='center', fig.height=7, fig.width=13}
mergeThresh <- 1 - 0.8
MEs <- moduleEigengenes(t(assays(se.expr.filt)$rpkm), colors=dynCol)$eigengenes
disMEs <- 1 - cor(MEs)
MEtree <- hclust(as.dist(disMEs), method='average')
plot(MEtree, main="Module Eigengene Clustering", xlab="Modules", sub="")
abline(h=mergeThresh, col='red')
```
```{r}
mergeEigen <- mergeCloseModules(t(assays(se.expr.filt)$rpkm), dynCol, cutHeight=mergeThresh)
mergedCol <- mergeEigen$colors
mergedMEs <- mergeEigen$newMEs
table(mergedCol)
```

We can now plot the result of the merge
```{r fig.align='center', fig.width=12, fig.height=7}
plotDendroAndColors(dendro=geneTree, colors=cbind(dynCol, mergedCol), 
                    groupLabels=c("Dynamic Colors", "Merged Colors"),
                    dendroLabels=FALSE, main="Merged Module Dendrogram"
                    )
```

We now add the module assignments to our *se.expr.filt* object for safekeeping.
```{r}
rowData(se.expr.filt)$module <- mergedCol
```

We can also export each module and their members into separate text files.
```{r}
# Create directory for modules
dir.create("/home/rstudio/work/modules", showWarnings=F)
for (clr in unique(rowData(se.expr.filt)$module)) {
    fname <- paste0("/home/rstudio/work/modules/", clr, ".module")
    write.table(row.names(se.expr.filt)[rowData(se.expr.filt)$module==clr],
                file=fname, sep='\n', quote=F, row.names=F, col.names=F)
}
```

## Functional Enrichment Analysis
Functional enrichment analysis will be done with the *enrichR* package. Here, we will analyze each module and determine what are the overarching "theme" of the module; that is, if there an overrepresentation of some kind of functionality exhibited by each particular set of genes found in each module.  

*EnrichR* is a tool that takes a list of gene symbols, compares that list to a specified background database, and outputs the enrichment. In our case, we will be using the 'GO_Biological_Process_2017' database to query for Gene Ontology terms related to biological processes.
```{r}
# Create enrichment directory
dir.create("/home/rstudio/work/enrichment", showWarnings=F)
db <- 'GO_Biological_Process_2017'
# We can ignore the grey module since those genes are just unassigned nodes
for (fmod in list.files("/home/rstudio/work/modules", 
                        full.names=TRUE, 
                        pattern="*.module")
     ) {
    if (fmod == "/home/rstudio/work/modules/grey.module") { next }
    # Get module color
    ofilename <- paste("/home/rstudio/work/enrichment/", 
                       unlist(strsplit(fmod, "/"))[[6]], 
                       ".enrich", 
                       sep=""
                       )
    mod.members <- scan(fmod, what=character())
    mem.as.sym <- rowData(se.expr.filt)$gene_symbol[rowData(se.expr.filt)$ensembl_gene_id 
                                                    %in% mod.members]
    this.enrichment <- enrichr(as.character(mem.as.sym), databases=db)
    printEnrich(this.enrichment, columns=c(1, 2, 3, 4, 7, 9), file=ofilename)
}
```

**Question 5**
What does each column represent? What is the top GO term in the darkgreen module? What is its adjusted P-value?

**Answer**
Columns can be found by exploring the *this.experiment$GO_Biological_Process_2017* matrix; they are 'Term', 'Overlap', 'P.value', 'Adjusted.P.value', 'Z.score', and 'Genes'. The top GO term in the darkgreen module is nervous system development with an adjusted p-value of 2.227390e-24.

We can furthermore visualize the significant GO terms for each module. For example,
```{r fig.width=15}
dark.green.enrich <- read.csv(file = "/home/rstudio/work/enrichment/darkgreen.module.enrich",
                              skip = 1,
                              header = FALSE,
                              sep = '\t')
dg.labels <- dark.green.enrich$V1[order(dark.green.enrich$V4)]
dark.green.enrich$V4 <- -log10(dark.green.enrich$V4)
# adjusted pvalue is col 4
ggplot(dark.green.enrich, aes(x = reorder(V1, V4), y = V4)) + 
  geom_bar(stat = "identity") + 
  coord_flip() + 
  labs(x = "", y = "-log(P-Value)", 
       title = "GO Enrichment of DarkGreen Module") + 
  theme.standard
```


```{r}
sessionInfo()
```


# References
1.  Allen Institute for Brain Science (2018). *BrainSpan Atlas of the Developing Human Brain*. RNA-seq Gencode v10 summarized to genes. Available from https://brainspan.org
2.  Chen, E. Y., Tan, C. M., Kou, Y., Duan, Q., Wang, Z., Meirelles, G. V., … Ma’ayan, A. (2013). *Enrichr: interactive and collaborative HTML5 gene list enrichment analysis tool*. BMC Bioinformatics, 14, 128. http://doi.org/10.1186/1471-2105-14-128
3.  Langfelder, P., & Horvath, S. (2008). *WGCNA: an R package for weighted correlation network analysis*. BMC Bioinformatics, 9, 559. http://doi.org/10.1186/1471-2105-9-559
4.  Kuleshov, M. V., Jones, M. R., Rouillard, A. D., Fernandez, N. F., Duan, Q., Wang, Z., … Ma’ayan, A. (2016). *Enrichr: a comprehensive gene set enrichment analysis web server 2016 update*. Nucleic Acids Research, 44(Web Server issue), W90–W97. http://doi.org/10.1093/nar/gkw377
